lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs

```diff
@@ -3163,75 +3162,132 @@ instance Buildable a => Show (ShowBuildable a) where
             -- NOTE: genTxOut does not generate quantities larger than
             -- `maxBound :: Word64`, however users could supply these.
             -- We should ideally test what happens, and make it clear what code,
             -- if any, should validate.
-            o <- Compatibility.fromCardanoTxOut <$> genTxOut Cardano.AlonzoEra
-            return (Compatibility.fromCardanoTxIn $ fst i, o, Nothing)
+            o <- genTxOut Cardano.AlonzoEra
+            return (fst i, o)
         let redeemers = []
         return $ PartialTx
             tx
-            resolvedInputs
+            inputUTxO
             redeemers
-    shrink (PartialTx tx inputs redeemers) =
-        [ PartialTx tx inputs' redeemers
-        | inputs' <- shrinkInputs inputs
+    shrink (PartialTx tx inputUTxO redeemers) =
+        [ PartialTx tx inputUTxO' redeemers
+        | inputUTxO' <- shrinkInputResolution inputUTxO
         ] ++
         [ restrictResolution $ PartialTx
             tx'
-            inputs
+            inputUTxO
             redeemers
         | tx' <- shrinkTxAlonzo tx
         ]
-      where
-        shrinkInputs (i:ins) =
-            map (:ins) (shrink i) ++ map (i:) (shrinkInputs ins)
-        shrinkInputs [] =
-            []
 
 instance Arbitrary (PartialTx Cardano.BabbageEra) where
     arbitrary = do
         let era = BabbageEra
         tx <- genTxForBalancing era
         let (Cardano.Tx (Cardano.TxBody content) _) = tx
         let inputs = Cardano.txIns content
-        resolvedInputs <- forM inputs $ \i -> do
+        inputUTxO <- fmap (Cardano.UTxO . Map.fromList) . forM inputs $ \i -> do
             -- NOTE: genTxOut does not generate quantities larger than
             -- `maxBound :: Word64`, however users could supply these.
             -- We should ideally test what happens, and make it clear what code,
             -- if any, should validate.
-            o <- Compatibility.fromCardanoTxOut <$> genTxOut Cardano.BabbageEra
-            return (Compatibility.fromCardanoTxIn $ fst i, o, Nothing)
+            o <- genTxOut Cardano.BabbageEra
+            return (fst i, o)
         let redeemers = []
-        return $ PartialTx tx resolvedInputs redeemers
-    shrink (PartialTx tx inputs redeemers) =
-        [ PartialTx tx inputs' redeemers | inputs' <- shrinkInputs inputs ] <>
-        [ restrictResolution $ PartialTx tx' inputs redeemers
-        | tx' <- shrinkTxBabbage tx ]
-      where
-        shrinkInputs (i:ins) =
-            map (:ins) (shrink i) ++ map (i:) (shrinkInputs ins)
-        shrinkInputs [] = []
+        return $ PartialTx tx inputUTxO redeemers
+    shrink (PartialTx tx inputUTxO redeemers) =
+        [ PartialTx tx inputUTxO' redeemers
+        | inputUTxO' <- shrinkInputResolution inputUTxO
+        ] <>
+        [ restrictResolution $ PartialTx tx' inputUTxO redeemers
+        | tx' <- shrinkTxBabbage tx
+        ]
```

Comment:
I tried making the balanceTx validity property use `BabbageEra` instead to be able to remove the `Alonzo` arbitrary instance, thinking while we'd ideally test both, this would be better. However, I ran into problems with keeping generators good enough for all `prop_balanceTransactionExistingReturnCollateral`,  `prop_balanceTransactionExistingTotalCollateral` and `prop_balanceTransactionValid`. I'd like to deal with this more properly so I'm leaving things as they are (with duplication) for now. (Btw, I'd like to remove the collateral properties in favour for a new check in `prop_balanceTransactionValid`.)